### ISSUE TITLE
Short taps on wakeup are interpreted as long taps (kobo glo hd N437)

### ISSUE THREAD

**Issue by worldexplorer** [koreader/koreader#4431]
* KOReader version: v2018.12-31-gf6743a4_2018_12_28
* Device: Kobo Glo HD, model N437

#### Issue
after I wake up my kobo (that has been put to sleep with short press on physical button on top), any tap on the screen is considered as a long tap. I tap to turn the page and a dictionary window pops up with highlighted word where I tapped; closing this popup window sometimes succeeds, sometimes it doesn't; I can sometimes move this popup window with the dictionary, but how koreader interprets my taps (single / double / swipe / drag and drop) is completely messed; the real long tap is also considered as a long tap, and doing long tap to kobo doesn't make short taps work;

what doesn't help:
1) having a screen cleaned up via Screensaver > Use Message as screensaver; I was hoping that re-drawing the page on wake up will unfreeze the tap sensor;
2) I'm not sure about the book format displayed; I think PDFs are not suffering from this issue;
3) putting kobo to sleep, and waking it up while holding finger on the screen without motion; having two fingers tap without motion while kobo wakes up and cleans the screen by redrawing white to black and back;
4) upgrading KSM (I used kobo start menu 08 for 1.5 years, and 2 weeks ago I upgraded to kobo start menu 09);
5) none of the koreader updates during 2018 addressed this issue - I didn't try every dev version, but "just waiting for the fix" didn't help :) 

what helps:
1) putting kobo to sleep, and waking it up while moving finger on the screen;
2) of course, full restart (twice, because first time kobo boots into nickel, after a forced long press; second time it boots to Kobo Start Menu 9 after second forced long press);

suggestion: koreader needs to initialize/reset the digitizer somehow (via kernel driver call?); it definitely has something to do with hardware, sensors, debouncing (as a hardware phenomenon) and kobo specifics; may be nickel has some weird code executed on wakeup that is visible under disassembler;

extra observations:
1) nickel (a factory rakuten reader software) never has this issue; while in nickel, putting to sleep by short pressing device's button on top and waking it up by the second press - always makes taps correct (never my taps to turn page are "seen" by nickel as long taps);
2) plato doesn't seem to have this issue, but I didn't use it too much yet (inconvenient folder access, I can't find a book I put into 3/4-level folder structure; with hundreds of books I need a listing and an explorer-like interface, instead of tags-like);
3) KSM never had this misinterpretation of taps (probably because it always starts from "cold hardware" reset state);

#### Steps to reproduce
1) get a kobo glo hd (if nobody else complains then probably only a single device that I have is broken);
2) install ANY version of koreader (I was observing this issue for 1 year hoping it will be fixed soon);
3) download any EPUB from Wiki;
4) do normal reading, turn the pages, put kobo to sleep;
5) every 10-15 "wake-ups" you will see exactly the same symptoms I described in the ISSUE section above;

#### MY BIG THANKS FOR:
1) a better "reflow" in PDFs;
2) wifi permanent connection and auto-restore of wifi connection on wakeup;
3) dictionaries that I can download from koreader top menu (before, it was a nightmare);
4) detailed (marked as dangerous) settings where I can mention that landscape will turn counter-clockwise (I couldn't prove it is working though, but may be next release);
5) text editor (!!) where I can put notes;
6) CSS styles applied to all books from the Style Tweaks menu - THIS IS GOLDEN!! Thank you

#### comparing to nickel, koreader is a gem!

**Comment by NiLuJe** [comment]
Does this affect the very first touch input when koreader starts, too? Because that one is often wonky on my h20, FWIW...

**Comment by worldexplorer** [comment]
no, after first touch in koreader after koreader starts from KSM08/09 - was always perfect; short tap is (was) always short tap, and long tap was always long tap. it has something to do with going ON back from the sleep, and I observed the mess with taps only after I woke up the kobo glo hd from the regular koreader sleep;
- rebooting kobo glo hd and clicking "start koreader" in ksm 08/09 was always mess free;
- restarting koreader (after an update) was from a state with unmessed taps,
- exiting from koreader and launching it again from ksm was always perfect (short taps were always short taps)

**Comment by worldexplorer** [comment]
it looks like just having dictionaries INSTALLED - will make the short tap be considered as a long tap.

#### steps to reproduce:
copy the "tesseract" "learned data files" from here
https://github.com/tesseract-ocr/tesseract/wiki/Data-Files#data-files-for-version-304305
to
.adds/koreader/data/tesseract
via USB. open any PDF and try long tap. it works, the dictionaries are looked up and the translation shows up in a popup. now, play with the top button, putting kobo to sleep and waking it up. one time out of 10-15 it will for sure refuse to turn pages and will pop up a translation window and the word underneath will be selected;

#### TEMP FIX I APPLIED TO SEE IF THE ISSUE WILL PERSIST
I renamed these folders (and created the new ones with the same names, in order to keep the dictionaries on my device and avoid downloading next time):
.adds/koreader/data/dict
.adds/koreader/data/tessdata
into something unexpected by koreader so that koreader "looses the dictionaries"; now both
- OCR engine compains that I have to download dictionaries while reading PDFs;
- HTML rendering - based EPUBs / FB2s also complain that dictionaries are not installed so I have to download them;

#### My next steps:
for the next few days I will be keeping watching whether the short taps will become considered as long taps; I will report whether I will observe the mess; I have a strong feeling that having NO DICTIONARIES at all in sight will heal this issue;

happy new year!

**Comment by worldexplorer** [comment]
So far I've red two books since my last comment, and I confirm:

- YES having NO dictionaries in sight of koreader REMOVED the issue;
- my assumption that "there must be some kernel driver call on wake up to restore the digitizer" is wrong;
- the issue arises in koreader, not in the hardware;

#### my next suggestion
probably dictionaries initialization happens in the main thread and can be delayed and scheduled for 5 seconds later after the "normal" wakeup (this can be "quick and dirty" approach but that's what I would try first)

**Comment by Frenzie** [comment]
There's no such thing as dictionary initialization though. It's just a matter of asking the sdcv binary for some output when the user requests a dictionary lookup.

**Comment by poire-z** [comment]
I too have a GloHD, dictionaries and frontlight gesture crontroller enabled, but I experience none of what you do... And reading what you observed and described, and the code, nothing strikes me.

What I sometimes get (once or twice a month) is that, suddenly while reading, tap and other screen gestures do not work, while the power button still do (so, I guess it's not KOReader being stuck in an infinite loop, just the touch hardware layer not working or (*)). So, I have to hold it to go to suspend, and wait 15 seconds for a real suspend to happen - then press it again for a real resume to clean things up, and it's all fine again (if I don't wait the 15 seconds, it's not a real suspend nor real resume, and touch is still not working).

(*) and it happens even more rarely that after a few tap and wait, it suddenly works again, while still being laggy, and that i'm in some kind of Hold situation, when my tap or taps (made while the kobo looked unresponsibe - I'm sure I did not make any long tap) has selected some text from some random start position... 
Because of that, I'm not 100% sure it's just the hardware touch layer. It could possibly be some touch event timestamp messup, like events being discarded because the timestamp is not right, or it's not the right type of event in the current context, until the moment some timestamp gets fine, or some delay has elapsed...  Obviously not easy to debug as it happens so rarely and can't be reproduced.
(pinging @NiLuJe just in case that blahblah rings a bell)

Regarding:

> after some OTA update (2-3 week of 2019), first filesystem read takes LONGER THAN in 2018 koreader versions; on wakeup, koreader opens my last book, but when I click on the FILES icon from the MENU, I have to wait 3-4 seconds for the files in the current folder 

That's explainable if that book was an EPUB: with the OTA upgrade, you probably got a version with some crengine update where the crengine cache file format version was incremented: the previous cache was invalidated, the whole book has to be re-parsed (so, could take from a few seconds to > 1mn with big books), and when you quit the book, that cache has to be written back (could take more than 10s with a big book). On subsequente opening, cache is re-used, opening is quick, leaving is quick too as the cache don't need to be rewritten - but if you change font size, you'll see it taking some time, and when leaving, it will take again some time to write the cache as rendering has changed.

**Comment by NiLuJe** [comment]
IIRC, the GloHD was the final Kobo device with an IR grid, right?

The only quirk I'm aware about it in terms of HW is the resume issue, and AFAIK it only ever affected the H2O, and we work-it-around.

Looking at an input dump for an (UP TO DATE, might be important) Glo HD, you *may* want to try replacing the `touch_alyssum_protocol = true,` quirk by the `touch_probe_ev_epoch_time = true,` one.
AFAICT, the TRACKING_ID is sane, but the timestamps aren't (The alyssum quirk appears to enforce timestamp trickery, which means that might be the case for other Mk. 6 devices where they should only be using the timestamp quirk, and not alyssum).

**Comment by NiLuJe** [comment]
Speaking of the IR suspend issue, does `/sys/devices/virtual/input/input1/neocmd` still exist on your devices?

**Comment by poire-z** [comment]
> does /sys/devices/virtual/input/input1/neocmd still exist on your devices?

I have the feeling I've seen it in the past (fw 3.14 for a long time - or may be I've just seen it in the code :), but I don't see it on fw3.19.5761 ...

My Glo HD is usable with:
```diff
 -- Kobo Glo HD:
 local KoboAlyssum = Kobo:new{
     model = "Kobo_alyssum",
     hasFrontlight = yes,
     touch_phoenix_protocol = true,
-    touch_alyssum_protocol = true,
+    -- touch_alyssum_protocol = true,
+    touch_probe_ev_epoch_time = true,
     display_dpi = 300,
 }
```
except that multi-touch doesn't work anymore (2 fingers swipe left or right for Toc or Bookmarks, diag swipe for zooming images...).
Adding `if ev.type == EV_ABS and ev.code == ABS_MT_TRACKING_ID then ev.value = ev.value - 1 end` to `probeEvEpochTime` does not help :|

(Unrelated, just for info: I have some "standby" state in my Koreader mods - a tap on my bottom right corner switch off the frontlight and do `echo standby > /sys/power/state` , while not touching `/sys/power/state-extended`. This allows the IR grid to not sleep and catch a tap and resume the device and continue what's after that `echo`. I haven't done any real battery % measurement, but I get hours in this standby state without a drop of battery %... are those permanent IR rays that cheap?!)

**Comment by Frenzie** [comment]
Can't be too much or they wouldn't stick 'em in bathrooms everywhere instead of just leaving the light on. :-P

**Comment by NiLuJe** [comment]
Huh.

Okay, that possibly makes sense, the log I had was for a single touch gesture (i.e., a tap) ;).

So, nope, no other ideas ;/.

**Comment by worldexplorer** [comment]
actually I always kept "Delay when exit(ing) from screensaver" = Disable(d)

this current issue #4431 in this thread might be resolved by another fix here (thanks for that one!)
https://github.com/koreader/koreader/issues/4624
because it deals with wakeup state even in its smaller setRotation()/close()/"full" piece;

After #4624 got fixed, I turned ON my other suggestions: frontlight controller and dictionaries;
I'll be watching for some time and I feel this current #4431 might get resolved auto-magically :)

**Comment by pazos** [comment]
@worldexplorer: please reopen this if you find this issue again, thanks!

**Comment by wvhulle** [comment]
I still have this problem on a KOBO Aura One with dictionaries and protection cover installed. Could someone explain what the eventual fix was?

**Comment by pazos** [comment]
@wvhulle: AFAICT I never had this issue in a KA1. Please open a new ticket for that with verbose logs attached (plugins -> 2nd page -> developer options -> enable debug logging / enable verbose debug logging).

Instead of attaching the entire log just reproduce the issue you have and copy only that part of the log (better if you can reproduce a few times in a row).

In my experience the KA1 has the wonkiest touchscreen of every eink devices but never faced such an issue.

**Comment by worldexplorer** [comment]
it wasn't fixed. i was still experiencing this issue until I gave that Kobo GLO HD away to my friend :) it was July 2019 and koreder was already slowly fading in and out (this kind of "delayed initialization" of the touchscreen / digitizer could have helped but it didn't). nikel had the same gentle fade-in, and never had a single-touch-interpreted-as-double-touch issue on the same device (that kobo glo hd that I gave away and koreader had this issue on it). my new kobo clara is a piece of art :)
